support 'datetime64' and 'timedelta64' types

ianna opened the pull request:
issue https://github.com/scikit-hep/awkward-1.0/issues/367

ianna committed:
start working on datetime and timedelta types

ianna committed:
add a test

ianna committed:
rename the test

ianna committed:
add datetime units to array metadata to interprete the data correctly

ianna committed:
add the number of base units in a step

ianna committed:
add datetime support stubs in reducers

ianna committed:
take Jim's suggestions

ianna commented on tests/test_0835-datetime-type.py:
```diff
@@ -0,0 +1,63 @@
+    array = ak.Array(numpy_array)
+    assert (
+        str(array.type)
+        == '3 * datetime64[parameters={"__array__": "datetime64", "__datetime64_data__": "<M8[s]", "__datetime64_unit__": "1s"}]'
```
the type is correct, I'm not sure if all these parameters are useful.

ianna commented on tests/test_0835-datetime-type.py:
```diff
@@ -0,0 +1,63 @@
+    assert array.tolist() == [
+        1595846471,
+        1546300800,
+        1577836800,
```
these are the number of `datetime64` units: seconds in this case. The units and the step are stored in ` __datetime64_unit__` parameter.

ianna replied:
another approach is to list them as `np.datetime64`:
```python
assert array.tolist() == [
         np.datetime64("2020-07-27T10:41:11"),
         np.datetime64("2019-01-01T00:00:00"),
         np.datetime64("2020-01-01T00:00:00"),
     ]
```

jpivarski replied:
It would be better if `tolist` makes a list of `np.datetime64` objects, but that would be automatic if `one_dimensional[0]` returns a `np.datetime64`. The `tolist` uses a recursive function defined in src/python/content.cpp that just calls `getitem_at` with appropriate `box`ing.

ianna replied:
yes, that's what is done in the last commit, I think :-)

ianna reviewed (commented):
@jpivarski - I'm not sure if I should box it differently or leave it as `uint64_t`

ianna committed:
convert datetime to string in to_list

ianna committed:
fix for failing test - check that dtype attribute is present

ianna committed:
try to box it as np.datetime

ianna committed:
start on reducers: min and max

ianna committed:
remove parameters, use format instead, add datetime64 array builder

ianna committed:
commit all before updating to master

ianna committed:
cleanup errors and remove commented out lines

ianna committed:
fix a typo in an enum type indexing

ianna committed:
add more tests

ianna committed:
fix tests

ianna committed:
add more reducer tests

ianna committed:
add test for timdedelta64 array builder

ianna committed:
ak.sum makes sense for timedelta64, but not for datetim64
ak.prod doesn't make sense for either - throw a value error

ianna committed:
format

ianna committed:
add datetime to highlevel

ianna reviewed (commented):
there are two issues to fix: 

- [ ] precision loss: for example, when a `np.datetime64("2020-05")` is converted to seconds `np.datetime64("2020-05-01T20:56:24.000000")`
- [x] build `timedelta64` for a high-level `ak.ArrayBuilder`

ianna committed:
add highlevel tests

jpivarski commented on src/awkward/_util.py:
```diff
@@ -490,7 +490,32 @@ def completely_flatten(array):
+                    )
+                    for x in array
+                ],
+            )
```
This is a Python iteration over the whole array, not vectorized: `for x in array` where `isinstance(array, ak.layout.NumpyArray)`. This would be a serious performance bomb.

I just checked in the src/python/content.cpp and there's no NumpyArray method that converts it into an `np.ndarray`, circumventing the buffer protocol, following `to_cupy` and `to_jax`. That's the function this would need to call: (1) check to see that the `array.format.startswith("M")` (or `"m"`) and (2) call the special method instead of just `asarray`.

ianna replied:
done, thanks!

jpivarski commented on src/awkward/operations/convert.py:
```diff
@@ -958,7 +967,24 @@ def to_list(array):
+                    array.format[array.format.index("[") + 1 : array.format.index("]")],
+                )
+                for x in array
+            ]
```
Same thing again: this is not only iterating over the array in Python, but constructing each NumPy object from a type that is parsed for every entry in the array. It needs to be a zero-copy wrapping.

ianna replied:
done, thanks!

jpivarski commented on tests/test_0835-datetime-type.py:
```diff
@@ -0,0 +1,408 @@
+    dt1 = np.datetime64("2020-03-27T10:41", "15s")
+    dt2 = np.datetime64("2020-05")
+    # FIXME: do we need to support this?
+    # builder.datetime64(dt.astype(np.int64), "datetime64[s]")
```
No, if someone passes an integer to `builder.datetime64`, they shouldn't expect it to work.

ianna replied:
ok, removed

jpivarski commented on tests/test_0835-datetime-type.py:
```diff
@@ -0,0 +1,408 @@
+    dt1 = np.datetime64("2020-03-27T10:41", "15s")
+    dt2 = np.datetime64("2020-05")
+    # FIXME: do we need to support this?
+    # builder.datetime64(dt.astype(np.int64), "datetime64[s]")
```
No (see above).

ianna replied:
ok, removed

jpivarski committed:
Added some failing tests.

jpivarski reviewed (changes requested):
I tried it out and ran into a few bugs, most likely minor, and added them to the testing suite as `test_more`.

I also made a modification to the repr format:

```
<Array [2021-06-03T10:00, ... 2021-06-03T13:00] type='4 * datetime64'>
```

and

```
<Array [60 minutes, 60 minutes, 60 minutes] type='3 * timedelta64'>
```

The string representation of datetime/timedelta types don't agree with [Datashape](https://datashape.readthedocs.io/en/latest/types.html#date-time-and-datetime), though. Is it possible to fix this? For one thing, the type strings should drop the "64", but is there a way to get timezone information in datetime (e.g. `datetime[tz='UTC']`)?

There is one significant problem, though: whenever we get data from a datetime/timedelta NumpyArray into an `np.ndarray`, it's through an expensive-looking loop in Python. This should be a zero-copy view. The NumpyArray class already has methods like `to_cupy` and `to_jax` that circumvent the buffer interface (and have to be called explicitly by the code layer above it): we need a similar one for `to_numpy` that would be called when the dtype is not something that can be expressed in a buffer.

Actually, we can short-cut this: I created a `view_int64` property on NumpyArray, which just drops the datetime/timedelta metadata so that you can send it through a buffer and add the metadata back in:

```python
>>> akarray
<Array [2021-06-03T10:00, 2021-06-03T11:00] type='2 * datetime64'>
>>> akarray.layout
<NumpyArray format="M8[m]" shape="2" data="0x 78ad9c01 00000000 b4ad9c01 00000000" at="0x562eaa6b1e10"/>
>>> akarray.layout.view_int64
<NumpyArray format="l" shape="2" data="27045240 27045300" at="0x562eaa6b1e10"/>
>>> np.asarray(akarray.layout.view_int64)
array([27045240, 27045300])
>>> np.asarray(akarray.layout.view_int64).view(akarray.layout.format)
array(['2021-06-03T10:00', '2021-06-03T11:00'], dtype='datetime64[m]')
```

Use this to fix the two Python loops over array data.

ianna committed:
add comments

ianna committed:
fix the performance bomb

ianna committed:
fix another one

ianna committed:
remove comments

ianna committed:
fix failing in python 2.7 tests

ianna committed:
bypass buffer issues with an M/m format in a few places, use an int64 view

ianna committed:
format

ianna committed:
reorder type check in builder_fromiter:
if py::isinstance(obj, py::module::import("numpy").attr("datetime64")) is true,
then py::isinstance(obj, py::module::import("numpy").attr("integer")) is also true

ianna committed:
do not run the test with numpy 1.13

jpivarski commented on src/awkward/operations/convert.py:
```diff
@@ -232,6 +232,12 @@ def to_numpy(array, allow_missing=True):
+        str(ak.operations.describe.type(array)) == "datetime64"
+        or str(ak.operations.describe.type(array)) == "timedelta64"
+    ):
+        return numpy.array([array[i] for i in range(len(array))])
```
`numpy.array([array[i] for i in range(len(array))])` in vectorized form is `numpy.arange(len(array))` (and you have to specify the dtype for consistency on Windows).

But why is `ak.to_numpy` returning `[0, 1, 2, 3, ...]` for datetimes/timedeltas?

jpivarski replied:
Also, if you fix the string representation ("datetime" and "timedelta" are the Datashape words; they leave off the "64"), then the type-check will be broken: check the format for `startswith("m")`.

ianna replied:
done

jpivarski commented on include/awkward/builder/Builder.h:
```diff
@@ -78,6 +78,14 @@ namespace awkward {
+    /// @brief Adds a timedelta64 value `x` to the accumulated data.
+    virtual const BuilderPtr
+      timedelta64(int64_t x, const std::string& unit) = 0;
+
```
For consistency, drop the "64" in these method names...

ianna replied:
done

ianna committed:
rename datetime64 to datetime and timedelta64 to timedelta

ianna committed:
cleanup

ianna committed:
not all arrays have a format attribute - revert to the check on type, do not manipulate with numpy array, return it asis

ianna committed:
format it

ianna committed:
tostring nicer output

ianna committed:
tidy up

ianna committed:
remove remaining 64 from reducer function names
eePlease enter the commit message for your changes. Lines starting

ianna committed:
format

ianna committed:
add pandas test

ianna committed:
fix for datetime from pandas

ianna commented on src/awkward/operations/convert.py:
```diff
@@ -970,10 +970,16 @@ def to_list(array):
-                .asarray(array.view_int64)
-                .view(array.format)
-                .tolist()
+                [
```
@jpivarski - couldn't figure out why `tolist` returns integers for an array with `M` format...

drahnreb replied:
This is somewhat expected behavior when providing `ns` resolution. 

```python
>>> np.datetime64('2019-09-02T09:30:00', 'ns').tolist()
1567416600000000000
>>> np.datetime64("2019-09-02T09:30:00.0000000").tolist()
1567416600000000000
```

It's a very [old and complex (if not weird)](https://stackoverflow.com/questions/13703720/converting-between-datetime-timestamp-and-datetime64/21916253#21916253) behavior.
Numpy considers datetime still [experimenting](https://numpy.org/doc/stable/reference/arrays.datetime.html), and this might change anytime, but I would prefer to mimic rather than sacrificing non-vectorized performance.

ianna replied:
thanks! fixed in https://github.com/scikit-hep/awkward-1.0/pull/911

ianna commented on src/python/content.cpp:
```diff
@@ -2313,14 +2285,22 @@ make_NumpyArray(const py::handle& m, const std::string& name) {
         py::array array = anyarray.cast<py::array>();
 
+        if (!PyObject_CheckBuffer(anyarray.ptr())) {
+          // anyarray does not support buffer protocol
```
@jpivarski - this catches a `pandas` case as well as a `builtin` case beneath:
```python
array = ak.layout.NumpyArray(
        [
            np.datetime64("2019-09-02T09:30:00"),
            np.datetime64("2019-09-13T09:30:00"),
            np.datetime64("2019-09-21T20:00:00"),
        ]
    )

```

ianna reviewed (commented):
@jpivarski - please, have a look when you have time. thanks!

jpivarski committed:
Fix ak.from_arrow and unit-scaling (should be floating point) to address last comment.

jpivarski committed:
More complete Arrow coverage and testing.
